### ISSUE TITLE
Compiler ICE

### ISSUE THREAD

**Issue by makspll** [rust-lang/rust-clippy#12639]
### Summary

ICE when running clippy/test/check/doc as seen in CI run:

https://github.com/makspll/bevy_mod_scripting/actions/runs/8580580540/job/23516782014?pr=112

### Version
Can reproduce locally with:
```text
rustc 1.77.1 (7cf61ebde 2024-03-27)
binary: rustc
commit-hash: 7cf61ebde7b22796c69757901dd346d0fe70bd97
commit-date: 2024-03-27
host: x86_64-unknown-linux-gnu
release: 1.77.1
LLVM version: 17.0.6
```

### Error output

<details><summary>Backtrace</summary>
  <p>

  ```
 thread 'rustc' panicked at compiler/rustc_hir_typeck/src/demand.rs:403:52:
index out of bounds: the len is 1 but the index is 1
stack backtrace:
   0:     0x7f4c6c18bd16 - std::backtrace_rs::backtrace::libunwind::trace::h6e4a662bea54ccfc
                               at /rustc/7cf61ebde7b22796c69757901dd346d0fe70bd97/library/std/src/../../backtrace/src/backtrace/libunwind.rs:104:5
   1:     0x7f4c6c18bd16 - std::backtrace_rs::backtrace::trace_unsynchronized::hb42b4eb2797d9c0e
                               at /rustc/7cf61ebde7b22796c69757901dd346d0fe70bd97/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7f4c6c18bd16 - std::sys_common::backtrace::_print_fmt::h2bc261f3223f4e4d
                               at /rustc/7cf61ebde7b22796c69757901dd346d0fe70bd97/library/std/src/sys_common/backtrace.rs:68:5
   3:     0x7f4c6c18bd16 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h9cca0343d66d16a8
                               at /rustc/7cf61ebde7b22796c69757901dd346d0fe70bd97/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x7f4c6c1de770 - core::fmt::rt::Argument::fmt::h8b666c45176be671
                               at /rustc/7cf61ebde7b22796c69757901dd346d0fe70bd97/library/core/src/fmt/rt.rs:142:9
   5:     0x7f4c6c1de770 - core::fmt::write::h4311bce0ee536615
                               at /rustc/7cf61ebde7b22796c69757901dd346d0fe70bd97/library/core/src/fmt/mod.rs:1120:17
   6:     0x7f4c6c17f58f - std::io::Write::write_fmt::h0685c51539d0a0cd
                               at /rustc/7cf61ebde7b22796c69757901dd346d0fe70bd97/library/std/src/io/mod.rs:1846:15
   7:     0x7f4c6c18baf4 - std::sys_common::backtrace::_print::h25f19b1d64e81f86
                               at /rustc/7cf61ebde7b22796c69757901dd346d0fe70bd97/library/std/src/sys_common/backtrace.rs:47:5
   8:     0x7f4c6c18baf4 - std::sys_common::backtrace::print::h2fb8f70628a241ed
                               at /rustc/7cf61ebde7b22796c69757901dd346d0fe70bd97/library/std/src/sys_common/backtrace.rs:34:9
   9:     0x7f4c6c18e887 - std::panicking::default_hook::{{closure}}::h05093fe2e3ef454d
  10:     0x7f4c6c18e5e9 - std::panicking::default_hook::h5ac38aa38e0086d2
                               at /rustc/7cf61ebde7b22796c69757901dd346d0fe70bd97/library/std/src/panicking.rs:292:9
  11:     0x7f4c6edab4dc - std[6985fe5dee7179c0]::panicking::update_hook::<alloc[c678e180b9908eaa]::boxed::Box<rustc_driver_impl[57d634a88fe18246]::install_ice_hook::{closure#0}>>::{closure#0}
  12:     0x7f4c6c18efd6 - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::hce488f674cf5618d
                               at /rustc/7cf61ebde7b22796c69757901dd346d0fe70bd97/library/alloc/src/boxed.rs:2029:9
  13:     0x7f4c6c18efd6 - std::panicking::rust_panic_with_hook::hed79743dc8b4b969
                               at /rustc/7cf61ebde7b22796c69757901dd346d0fe70bd97/library/std/src/panicking.rs:785:13
  14:     0x7f4c6c18ed22 - std::panicking::begin_panic_handler::{{closure}}::ha437b5d58f431abf
                               at /rustc/7cf61ebde7b22796c69757901dd346d0fe70bd97/library/std/src/panicking.rs:659:13
  15:     0x7f4c6c18c216 - std::sys_common::backtrace::__rust_end_short_backtrace::hd98e82d5b39ec859
                               at /rustc/7cf61ebde7b22796c69757901dd346d0fe70bd97/library/std/src/sys_common/backtrace.rs:171:18
  16:     0x7f4c6c18ea74 - rust_begin_unwind
                               at /rustc/7cf61ebde7b22796c69757901dd346d0fe70bd97/library/std/src/panicking.rs:647:5
  17:     0x7f4c6c1dae75 - core::panicking::panic_fmt::hc69c4d258fe11477
                               at /rustc/7cf61ebde7b22796c69757901dd346d0fe70bd97/library/core/src/panicking.rs:72:14
  18:     0x7f4c6c1db0b2 - core::panicking::panic_bounds_check::he5254f424ac3a4c4
                               at /rustc/7cf61ebde7b22796c69757901dd346d0fe70bd97/library/core/src/panicking.rs:208:5
  19:     0x7f4c6ef18f25 - <core[a93eb12aade15298]::slice::iter::Iter<&rustc_hir[9e01f25f0b9ccddd]::hir::Expr> as core[a93eb12aade15298]::iter::traits::iterator::Iterator>::find_map::<rustc_middle[42dfbbbcfa5f0e07]::ty::Ty, <rustc_hir_typeck[80f3f778ae8aa14c]::fn_ctxt::FnCtxt>::note_source_of_type_mismatch_constraint::{closure#3}>
  20:     0x7f4c6ef5e326 - <rustc_hir_typeck[80f3f778ae8aa14c]::fn_ctxt::FnCtxt>::note_source_of_type_mismatch_constraint
  21:     0x7f4c6ef75a2d - <rustc_hir_typeck[80f3f778ae8aa14c]::fn_ctxt::FnCtxt>::report_arg_errors
  22:     0x7f4c70fddaf7 - <rustc_hir_typeck[80f3f778ae8aa14c]::fn_ctxt::FnCtxt>::check_method_argument_types
  23:     0x7f4c70a80ac0 - <rustc_hir_typeck[80f3f778ae8aa14c]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  24:     0x7f4c704962aa - <rustc_hir_typeck[80f3f778ae8aa14c]::fn_ctxt::FnCtxt>::check_block_with_expected
  25:     0x7f4c70a806a2 - <rustc_hir_typeck[80f3f778ae8aa14c]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  26:     0x7f4c7049929a - <rustc_hir_typeck[80f3f778ae8aa14c]::fn_ctxt::FnCtxt>::check_decl
  27:     0x7f4c704971c8 - <rustc_hir_typeck[80f3f778ae8aa14c]::fn_ctxt::FnCtxt>::check_block_with_expected
  28:     0x7f4c70a806a2 - <rustc_hir_typeck[80f3f778ae8aa14c]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  29:     0x7f4c7079c6ce - rustc_hir_typeck[80f3f778ae8aa14c]::check::check_fn
  30:     0x7f4c70a73297 - rustc_hir_typeck[80f3f778ae8aa14c]::typeck
  31:     0x7f4c70a725b9 - rustc_query_impl[cc38f1a9c625188c]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[cc38f1a9c625188c]::query_impl::typeck::dynamic_query::{closure#2}::{closure#0}, rustc_middle[42dfbbbcfa5f0e07]::query::erase::Erased<[u8; 8usize]>>
  32:     0x7f4c702e8862 - rustc_query_system[da4130aa38fb7c58]::query::plumbing::try_execute_query::<rustc_query_impl[cc38f1a9c625188c]::DynamicConfig<rustc_query_system[da4130aa38fb7c58]::query::caches::VecCache<rustc_span[e4948e3682155d21]::def_id::LocalDefId, rustc_middle[42dfbbbcfa5f0e07]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[cc38f1a9c625188c]::plumbing::QueryCtxt, false>
  33:     0x7f4c702e8490 - rustc_query_impl[cc38f1a9c625188c]::query_impl::typeck::get_query_non_incr::__rust_end_short_backtrace
  34:     0x7f4c702e6437 - rustc_hir_analysis[8407a811d94540bd]::check_crate
  35:     0x7f4c70a60e66 - rustc_interface[be390c89f505a55d]::passes::analysis
  36:     0x7f4c70a6085f - rustc_query_impl[cc38f1a9c625188c]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[cc38f1a9c625188c]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[42dfbbbcfa5f0e07]::query::erase::Erased<[u8; 1usize]>>
  37:     0x7f4c70a082f4 - rustc_query_system[da4130aa38fb7c58]::query::plumbing::try_execute_query::<rustc_query_impl[cc38f1a9c625188c]::DynamicConfig<rustc_query_system[da4130aa38fb7c58]::query::caches::SingleCache<rustc_middle[42dfbbbcfa5f0e07]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[cc38f1a9c625188c]::plumbing::QueryCtxt, false>
  38:     0x7f4c70a08139 - rustc_query_impl[cc38f1a9c625188c]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  39:     0x7f4c70e8cecb - rustc_interface[be390c89f505a55d]::interface::run_compiler::<core[a93eb12aade15298]::result::Result<(), rustc_span[e4948e3682155d21]::ErrorGuaranteed>, rustc_driver_impl[57d634a88fe18246]::run_compiler::{closure#0}>::{closure#0}
  40:     0x7f4c70f34a1f - std[6985fe5dee7179c0]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[be390c89f505a55d]::util::run_in_thread_with_globals<rustc_interface[be390c89f505a55d]::interface::run_compiler<core[a93eb12aade15298]::result::Result<(), rustc_span[e4948e3682155d21]::ErrorGuaranteed>, rustc_driver_impl[57d634a88fe18246]::run_compiler::{closure#0}>::{closure#0}, core[a93eb12aade15298]::result::Result<(), rustc_span[e4948e3682155d21]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[a93eb12aade15298]::result::Result<(), rustc_span[e4948e3682155d21]::ErrorGuaranteed>>
  41:     0x7f4c70f3487d - <<std[6985fe5dee7179c0]::thread::Builder>::spawn_unchecked_<rustc_interface[be390c89f505a55d]::util::run_in_thread_with_globals<rustc_interface[be390c89f505a55d]::interface::run_compiler<core[a93eb12aade15298]::result::Result<(), rustc_span[e4948e3682155d21]::ErrorGuaranteed>, rustc_driver_impl[57d634a88fe18246]::run_compiler::{closure#0}>::{closure#0}, core[a93eb12aade15298]::result::Result<(), rustc_span[e4948e3682155d21]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[a93eb12aade15298]::result::Result<(), rustc_span[e4948e3682155d21]::ErrorGuaranteed>>::{closure#1} as core[a93eb12aade15298]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  42:     0x7f4c6c198675 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h32ae492e80523c39
                               at /rustc/7cf61ebde7b22796c69757901dd346d0fe70bd97/library/alloc/src/boxed.rs:2015:9
  43:     0x7f4c6c198675 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::hd05b2dc112b7a972
                               at /rustc/7cf61ebde7b22796c69757901dd346d0fe70bd97/library/alloc/src/boxed.rs:2015:9
  44:     0x7f4c6c198675 - std::sys::pal::unix::thread::Thread::new::thread_start::h40e6fd3f8ce15a14
                               at /rustc/7cf61ebde7b22796c69757901dd346d0fe70bd97/library/std/src/sys/pal/unix/thread.rs:108:17
  45:     0x7f4c6bc94ac3 - <unknown>
  46:     0x7f4c6bd26850 - <unknown>
  47:                0x0 - <unknown>

error: the compiler unexpectedly panicked. this is a bug.
Error: the compiler unexpectedly panicked. this is a bug.
note: we would appreciate a bug report: https://github.com/rust-lang/rust-clippy/issues/new?template=ice.yml

note: rustc 1.77.1 (7cf61ebde 2024-03-27) running on x86_64-unknown-linux-gnu

note: compiler flags: --crate-type lib -C opt-level=2 -C embed-bitcode=no -C codegen-units=8 -C debug-assertions=on -C strip=debuginfo

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
#0 [typeck] type-checking `bindings::<impl at crates/bevy_mod_scripting_core/src/bindings.rs:36:1: 36:22>::reflect_mut`
#1 [analysis] running analysis passes on this crate
end of query stack
note: Clippy version: clippy 0.1.77 (7cf61eb 2024-03-27)

error: could not compile `bevy_mod_scripting_core` (lib)
Error: The process '/home/runner/.cargo/bin/cargo' failed with exit code 101
  ```

  </p>
</details>

**Comment by matthiaskrgr** [comment]
This is probably the same as https://github.com/rust-lang/rust/issues/118185

**Comment by blyxyas** [comment]
I can confirm the last comment, this should be fixed as soon as rust-lang/rust#118185 gets a fix (there's already a few eyes on that issue, so I don't think it will take much, I'll still look at it and determine if I can get any new 

Thanks for the report! I'll close it for now because we can't really do anything :) You can subscribe to that issue if you want to be notified about future changes :heart:
